torch/distributed/_shard/sharded_tensor/_ops/math_ops.py

```diff
@@ -0,0 +1,75 @@
+import torch
+from torch import Tensor
+from torch.distributed._shard.sharded_tensor import (
+    ShardedTensor,
+    sharded_op_impl
+)
+from torch.distributed._shard.sharding_spec import ChunkShardingSpec
+from torch.distributed._shard.replicated_tensor import ReplicatedTensor
+
+def register_math_op(op):
+    @sharded_op_impl(op)
+    def binary_math_op(types, args=(), kwargs=None, pg=None):
+        """
+        Handles ``__torch_function__`` dispatch for the elementwise op such
+        as ``torch.nn.functional.gelu`` or ``torch.nn.functional.relu``.
+        This method computes on either a normal tensor or a sharded tensor.
+        """
+        if len(args) != 2:
+            raise ValueError("Only support binary math op on ShardedTensor for now!")
+        lhs = args[0]
+        rhs = args[1]
+        # Validate types
+        if isinstance(lhs, ShardedTensor) and isinstance(rhs, ShardedTensor):
+            lhs_spec = lhs.sharding_spec()
+            rhs_spec = rhs.sharding_spec()
+            if not isinstance(lhs_spec, ChunkShardingSpec) or not isinstance(rhs_spec, ChunkShardingSpec):
+                raise TypeError("Only ShardedTensor with ChunkShardingSpec supports"
+                                " two ShardedTensor together")
+
+            if lhs.size() == rhs.size() and lhs_spec.sharding_dim == rhs_spec.sharding_dim:
+                # perform local element-wise math op
+                res = op(lhs.local_tensor(), rhs.local_tensor)
+                return ShardedTensor._init_from_local_tensor(res, lhs_spec, lhs.size(), process_group=pg)
+            else:
+                raise RuntimeError("Implicit broadcasting not supported yet!")
+
+        elif isinstance(lhs, ReplicatedTensor):
+            assert isinstance(rhs, ShardedTensor)
+            if lhs.size() == rhs.local_tensor().size():
+                res = op(lhs, rhs.local_tensor())
+                return ShardedTensor._init_from_local_tensor(res, rhs.sharding_spec(), rhs.size(), process_group=pg)
+            else:
+                raise RuntimeError("Implicit broadcasting not supported yet!")
+
+        elif isinstance(rhs, ReplicatedTensor):
+            assert isinstance(lhs, ShardedTensor)
+            if rhs.size() == lhs.local_tensor().size():
+                res = op(lhs.local_tensor(), rhs)
+                return ShardedTensor._init_from_local_tensor(res, lhs.sharding_spec(), lhs.size(), process_group=pg)
+            else:
+                raise RuntimeError("Implicit broadcasting not supported yet!")
+        else:
+            raise RuntimeError(
+                f"torch function '{op.__name__}', with args: {args} and "
+                f"kwargs: {kwargs} not supported yet for ShardedTensor!")
+
+binary_ops = [
+    # add
+    torch.add,
+    Tensor.add,
+    # sub
+    torch.sub,
+    Tensor.sub,
+    Tensor.__rsub__,
+    # mul
+    torch.mul,
+    Tensor.mul,
```

Comment:
Thanks for making this happens. Can you also kindly add `torch.baddbmm` and `torch.bmm` because Megatron-LM use it for Transformer model (https://github.com/NVIDIA/Megatron-LM/blob/e156d2fea7fc5c98e645f7742eb86b643956d840/megatron/model/transformer.py#L349). So I guess instead of creating a new PR, adding it here might be faster? If you feel this PR is too big, I can work on another PR on top of this one. Thanks!

Reply:
sure, let me try adding it in the PR

Reply:
@fduwjj @wanchaol Let us focus on the fairseq XLMR-XL ops first: https://github.com/pytorch/fairseq/blob/c71870f370455e6154c730e8822ea323b5f266f6/fairseq/modules/transformer_layer.py#L20. Also, note that `torch.baddbmm` and `torch.bmm` are not binary math ops. But rather matmul ops.

Lets address these in a separate PR

Reply:
yeah sounds good, I actually tried with `baddbmm` and `bmm`, it was a bit complicated as it involves communication (similar to linear). As pritam suggested, given that they are matmul ops and not binary ops, lets leave it out of this PR and we can address the XLMR-XL ops first @fduwjj

Reply:
Sure. But it looks like `bmm` is indeed used in the XLMR: https://github.com/pytorch/fairseq/blob/c71870f370455e6154c730e8822ea323b5f266f6/fairseq/modules/multihead_attention.py#L504. I think I can work on this on my end.